internal/devpkg/narinfo_cache.go

```diff
@@ -103,39 +72,30 @@ func FillNarInfoCache(ctx context.Context, packages ...*Package) error {
 
 	group, _ := errgroup.WithContext(ctx)
 	for _, p := range eligiblePackages {
-		// If the package's NarInfo status is already known, skip it
-		isNarInfoInCache.lock.RLock()
-		_, ok := isNarInfoInCache.status[p.Raw]
-		isNarInfoInCache.lock.RUnlock()
-		if ok {
-			continue
-		}
 		pkg := p // copy the loop variable since its used in a closure below
 		group.Go(func() error {
-			err := pkg.fillNarInfoCache()
-			if err != nil {
-				// default to false if there was an error, so we don't re-try
-				isNarInfoInCache.lock.Lock()
-				isNarInfoInCache.status[pkg.Raw] = false
-				isNarInfoInCache.lock.Unlock()
-			}
+			_, err := pkg.fillNarInfoCacheIfNeeded()
 			return err
 		})
 	}
 	return group.Wait()
 }
 
-// fillNarInfoCache fills the cache value for the narinfo of this package,
+// fillNarInfoCacheIfNeeded fills the cache value for the narinfo of this package,
 // assuming it is eligible for the binary cache. Callers are responsible
 // for checking isEligibleForBinaryCache before calling this function.
 //
 // NOTE: this must be concurrency safe.
-func (p *Package) fillNarInfoCache() error {
+func (p *Package) fillNarInfoCacheIfNeeded() (bool, error) {
+	if status, alreadySet := isNarInfoInCache.Load(p.Raw); alreadySet {
+		return status.(bool), nil
+	}
+	isNarInfoInCache.Store(p.Raw, false) // store in case of failure
```

Comment:
There could be a race here where multiple goroutines see `alreadySet == false` before reaching line 93. Or one goroutine might reach line 93 and then another sees `status.(bool) == false` on line 91.

Reply:
Good call. I think our calling patterns make this unlikely but better to avoid if possible.

I think using `defer` might solve most race conditions. It's not immune to a success following a failure (but neither is the current code so it would not be a regression)